New article for DateTime::Format::RelativeTime (#401)
perladvent/perldotcom: content/article/release-of-new-module-datetime-format-relativetime.md

{

>     "title"       : "Release of new module DateTime::Format::RelativeTime",

Reviewer: ```suggestion
    "title"       : "Introducing DateTime::Format::RelativeTime",
```

    "authors"     : ["jacques-deguest"],
    "date"        : "2025-01-04T12:29:19",
    "tags"        : ["datetime", "l10n", "unicode", "cldr"],
    "draft"       : false,
    "image"       : "/images/release-of-new-module-datetime-format-relativetime/perl-for-relative-time-formatting.jpg",
    "thumbnail"   : "",

>     "description" : "Release of the new Perl module DateTime::Format::RelativeTime, designed to mirror its equivalent Web API Intl.RelativeTimeFormat",

Reviewer: ```suggestion
    "description" : "Introducing the new Perl module DateTime::Format::RelativeTime, designed to mirror its equivalent Web API: Intl.RelativeTimeFormat",
```

However, beyond the standard options, and parameters you can pass to the methods `format` and `formatToParts` (or `format_to_parts` if you prefer), you can also provide 1 or 2 `DateTime` objects, and `DateTime::Format::RelativeTime` will figure out for you the greatest difference between the 2 objects.

If you provide only 1 `DateTime` object, `DateTime::Format::RelativeTime` will instantiate a second one with `DateTime->now` and using the first `DateTime` object `time_zone` value.

For example:

```perl
my $dt = DateTime->new(
    year => 2024,
    month => 8,
    day => 15,
);
$fmt->format( $dt );
# Assuming today is 2024-12-31, this would return: "1 qtr. ago"
```

or, with 2 `DateTime` objects:

```perl
my $dt = DateTime->new(
    year => 2024,
    month => 8,
    day => 15,
);
my $dt2 = DateTime->new(
    year => 2022,
    month => 2,
    day => 22,
);
$fmt->format( $dt => $dt2 ); # "2 yr. ago"
```

> When using the method `formatToParts` (or `format_to_parts`) you will receive an array reference of hash reference making it easy to customise and handle as you wish. For example:

Reviewer: ```suggestion
When using the method `formatToParts` (or `format_to_parts`) you will receive an array reference of hash references, making it easy to customise and handle as you wish. For example:
```

```perl
use DateTime::Format::RelativeTime;
use Data::Pretty qw( dump );
my $fmt = DateTime::Format::RelativeTime->new( 'en', { numeric => 'auto' });
my $parts = $fmt->formatToParts( 10, 'seconds' );
say dump( $parts );
```

would yield:

```perl
[
    { type => "literal", value => "in " },
    { type => "integer", unit => "second", value => 10 },
    { type => "literal", value => " seconds" },
]
```

> You can use negative number to indicate the past, and you can also use decimals, such as:

Reviewer: ```suggestion
You can use a negative number to indicate the past, and you can also use decimals, such as:
```

> ```

Reviewer: ```suggestion
```perl
```

my $parts = $fmt->formatToParts( -3, 'minutes' );
say dump( $parts );
```

would yield:

```perl
[
    { type => "literal", value => "قبل " },
    { type => "integer", value => '٣', unit => "minute" },
    { type => "literal", value => " دقائق" },
]
```

or, here we are explicitly setting the numbering system to `deva`, which is not a system default:

```perl
use DateTime::Format::RelativeTime;
use Data::Pretty qw( dump );
my $fmt = DateTime::Format::RelativeTime->new( 'hi-IN', { numeric => 'auto', numberingSystem => 'deva' });
my $parts = $fmt->formatToParts( -3.5, 'minutes' );
say dump( $parts );
```

would yield:

```perl
[
    { type => "integer", value => '३', unit => "minute" },
    { type => "decimal", value => ".", unit => "minute" },
    { type => "fraction", value => '५', unit => "minute" },
    { type => "literal", value => " मिनट पहले" },
]
```

The option `numeric` can be set to `auto` or `always`. If it is on `auto`, the API will check if it can find a time relative term, such as `today` or `yesterday` instead of returning `in 0 day` or `1 day ago`. If it is set to `always`, then the API will always return a format involving a number like the ones I just mentioned.

> Just like its Web API counterpart, you can use the method `resolvedOptions` after having instantiated an object to check how `DateTime::Format::RelativeTime` has resolved the options you provided, or possibly came up with default ones:

Reviewer: ```suggestion
Just like its Web API counterpart, you can use the method `resolvedOptions` after having instantiated an object to check how `DateTime::Format::RelativeTime` has resolved the options you provided, or possibly come up with default ones:
```

> ```

Reviewer: ```suggestion
```perl
```

use DateTime::Format::RelativeTime;
use Data::Pretty qw( dump );
# locale is "Marathi"
my $fmt = DateTime::Format::RelativeTime->new( 'mr', { numeric => 'auto' });
my $options = $fmt->resolvedOptions;
say dump( $options );
```

would yield:

> ```

Reviewer: ```suggestion
```perl
```
